File: 100dynamic/931.swift
class Solution {
    func minFallingPathSum(matrix: [[Int]]) -> Int {
        let n = matrix.count
        var dp = Array(repeating: 0, count: n)

        for i in 0..<n {
            dp[i] = matrix[0][i]
        }

        for i in 1..<n {
            var last = Array(repeating: 0, count: n)
            for j in 0..<n {
                if j == 0 {
                    last[j] = min(dp[j], dp[j+1])
                } else if j == n-1 {
                    last[j] = min(dp[j], dp[j-1])
                } else {
                    last[j] = min(dp[j], min(dp[j-1],dp[j+1]))
                }
                last[j] += matrix[i][j]
            }
            dp = last
        }
        return dp.min()!
    }
}

File: 100dynamic/931.py
class Solution:
    def minNum(self, x,y,z):
        tmp = min(x,y)
        return min(tmp,z)
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        n = len(matrix)
        dp = [0] * n

        for i in range(n):
            dp[i] = matrix[0][i]
        last = dp[:]
        for i in range(1,n):
            for j in range(n):
                if (j == 0):
                    dp[j] = min(last[j],last[j+1]) + matrix[i][j]
                elif (j == n-1):
                    dp[j] = min(last[j],last[j-1]) + matrix[i][j]
                else:
                    dp[j] = self.minNum(last[j],last[j-1],last[j+1]) + matrix[i][j]
            last = dp[:]
        return min(dp)